communicator/ssh/communicator.go

```diff
@@ -413,18 +413,22 @@ func (c *comm) sftpUploadFile(path string, input io.Reader, client *sftp.Client,
 
 	// find out if destination is a directory (this is to replicate rsync behavior)
 	testDirectoryCommand := fmt.Sprintf(`test -d "%s"`, path)
-	cmd := &packer.RemoteCmd{Command: testDirectoryCommand}
+
+	cmd := &packer.RemoteCmd{
+		Command: testDirectoryCommand,
+	}
 
 	err := c.Start(cmd)
 
 	if err != nil {
-		log.Printf("Unable to check whether remote path is a dir: %s", err)
+		log.Printf("[ERROR] Unable to check whether remote path is a dir: %s", err)
 		return err
 	}
 	cmd.Wait()
 	if cmd.ExitStatus == 0 {
-		log.Printf("path is a directory; copying file into directory.")
-		path = filepath.Join(path, filepath.Base((*fi).Name()))
+		return fmt.Errorf(
```

Comment:
I'm fine with this; it's explicit, which is the most important thing. I can change this for the other provisioners too, in a different PR, so they remain consistent.